internal/internal_worker_base.go

```diff
@@ -344,52 +414,79 @@ func (bw *baseWorker) isStop() bool {
 	}
 }
 
-func (bw *baseWorker) runPoller() {
+func (bw *baseWorker) runPoller(taskWorker scalableTaskPoller) {
 	defer bw.stopWG.Done()
+	// TODO(quinn): with poller autoscaling, this metric doesn't make a lot of sense
+	// since the number of pollers can go up and down.
 	bw.metricsHandler.Counter(metrics.PollerStartCounter).Inc(1)
 
 	ctx, cancelfn := context.WithCancel(context.Background())
 	defer cancelfn()
 	reserveChan := make(chan *SlotPermit)
 
 	for {
-		bw.stopWG.Add(1)
-		go func() {
-			defer bw.stopWG.Done()
-			s, err := bw.slotSupplier.ReserveSlot(ctx, &bw.options.slotReservationData)
-			if err != nil {
-				if !errors.Is(err, context.Canceled) {
-					bw.logger.Error(fmt.Sprintf("Error while trying to reserve slot: %v", err))
-					select {
-					case reserveChan <- nil:
-					case <-ctx.Done():
-						return
-					}
+		// TODO(quinn): is there a way we can refactor this to avoid the function?
+		if func() bool {
```

Comment:
You could set a var which is this fn's return value and `continue` if false, but it makes the defer stuff harder too... I think this is totally fine.